### PR TITLE
run hg binary only once per isWorking()

### PR DESCRIPTION
**Pull request by vladak**
This mimics the changes done in `GitRepository` to avoid checking Mercurial command for some of repository operations repeatedly. Should help repository scan/invalidation performance and thus web app startup.

### CODE REVIEW

File: opengrok-indexer/src/main/java/org/opengrok/indexer/history/MercurialRepository.java [hunk 1]
```diff
@@ -104,6 +105,12 @@
     private static final Pattern LOG_COPIES_PATTERN
             = Pattern.compile("^(\\d+):(.*)");
 
+    /**
+     * This is a static replacement for 'working' field. Effectively, check if hg is working once in a JVM
+     * instead of calling it for every MercurialRepository instance.
+     */
+    private static final LazilyInstantiate<Boolean> HG_IS_WORKING = LazilyInstantiate.using(MercurialRepository::isHgWorking);
```
Comment (ahornace) [inline opengrok-indexer/src/main/java/org/opengrok/indexer/history/MercurialRepository.java:112]: Just a note for future, I'd personally change the name of `LazilyInstantiate` to `MemoizingSupplier` which follows more common naming or even better use one from Guava https://guava.dev/releases/19.0/api/docs/com/google/common/base/Suppliers.html#memoize(com.google.common.base.Supplier)

**Review by ahornace — APPROVED** [review]
Good idea!

File: opengrok-indexer/src/main/java/org/opengrok/indexer/history/MercurialRepository.java (hunk 1 above)
Comment (ahornace) [reply opengrok-indexer/src/main/java/org/opengrok/indexer/history/MercurialRepository.java:112]: Also, I think using supplier type is nicer:
 `Supplier<Boolean> HG_IS_WORKING = ...`

Comment (vladak) [reply opengrok-indexer/src/main/java/org/opengrok/indexer/history/MercurialRepository.java]: ok, used Supplier.

**Comment by vladak** [comment]
thanks @ahornace !

**Pull request merged**